modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -0,0 +1,312 @@
+
+(def ^:private bigquery-scope
+  "The scope to use for executing BigQuery requests; see:
+  `https://cloud.google.com/bigquery/docs/samples/bigquery-auth-drive-scope`.
+  Unclear if this can be sourced from the `com.google.cloud.bigquery` package directly."
+  "https://www.googleapis.com/auth/bigquery")
+
+(defn- ^ServiceAccountCredentials database->service-account-credential
+  "Returns a `ServiceAccountCredentials` (not scoped) for the given DB, from its service account JSON."
+  ^ServiceAccountCredentials [{{:keys [^String service-account-json]} :details
+                               :as             db}]
+  {:pre [(map? db) (seq service-account-json)]}
+  (ServiceAccountCredentials/fromStream (ByteArrayInputStream. (.getBytes service-account-json))))
+
+(defn- ^BigQuery database->client
+  [database]
+  (let [creds   (database->service-account-credential database)
+        bq-bldr (doto (BigQueryOptions/newBuilder)
+                  (.setCredentials (.createScoped creds (Collections/singletonList bigquery-scope))))]
+    (.. bq-bldr build getService)))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                      Sync                                                      |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defn- ^Iterable list-tables
+  "Fetch all tables (new pages are loaded automatically by the API)."
+  ([{{:keys [project-id dataset-id]} :details, :as database}]
+   (let [bq (database->client database)]
+     (list-tables bq project-id dataset-id)))
+
+  #_([{{:keys [project-id dataset-id]} :details, :as database}, ^String page-token-or-nil]
+     ;; RAR 2020-May-11 - this gets messy, because this function is used to determine if we
+     ;; can connect to the database. If a service account JSON is being used, the project ID
+     ;; field isn't required in the admin page because it's embedded in that JSON. During the
+     ;; initial save, the database entry doesn't exist yet, so we haven't pulled the project ID
+     ;; from that JSON. In this case, we need to go look it up in the credentials.
+     ;; TODO: maybe handle this better in `driver/can-connect? :bigquery-cloud-sdk`
+     (let [bq (database->client database)]
+      (list-tables db-client project-id dataset-id page-token-or-nil)))
+
+  ([^BigQuery client, ^String project-id ^String dataset-id]
+   {:pre [client (not (str/blank? dataset-id))]}
+   (.iterateAll (.listTables client (DatasetId/of project-id dataset-id) (u/varargs BigQuery$TableListOption [])))))
+
+(defmethod driver/describe-database :bigquery-cloud-sdk
+  [_ database]
+  ;; first page through all the 50-table pages until we stop getting "next page tokens"
+  (let [tables (list-tables database) #_(loop [tables [], ^TableList table-list (list-tables database)]
```

Comment:
why is the commented out code still here?

Reply:
Oops, removed.